gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,398 @@
+  private CatalogName catalogName;
+  private BigLakeClient client;
+
+  // Must have a no-arg constructor to be dynamically loaded
+  // initialize(String name, Map<String, String> properties) will be called to complete
+  // initialization
+  public BigLakeCatalog() {}
+
+  @Override
+  public void initialize(String inputName, Map<String, String> properties) {
+    Preconditions.checkArgument(
+        properties.containsKey(PROPERTIES_KEY_GCP_PROJECT), "GCP project must be specified");
+    String propProjectId = properties.get(PROPERTIES_KEY_GCP_PROJECT);
+    String propLocation =
+        properties.getOrDefault(PROPERTIES_KEY_GCP_LOCATION, DEFAULT_GCP_LOCATION);
+
+    BigLakeClient newClient;
+    try {
+      // TODO: to add more auth options of the client. Currently it uses default auth
+      // (https://github.com/googleapis/google-cloud-java#application-default-credentials)
+      // that works on GCP services (e.g., GCE, GKE, Dataproc).
+      newClient =
+          new BigLakeClient(
+              properties.getOrDefault(
+                  PROPERTIES_KEY_BIGLAKE_ENDPOINT, DEFAULT_BIGLAKE_SERVICE_ENDPOINT),
+              propProjectId,
+              propLocation);
+    } catch (IOException e) {
+      throw new ServiceFailureException(e, "Creating BigLake client failed");
+    }
+    initialize(inputName, properties, propProjectId, propLocation, newClient);
+  }
+
+  @VisibleForTesting
+  void initialize(
+      String inputName,
+      Map<String, String> properties,
+      String propProjectId,
+      String propLocation,
+      BigLakeClient bigLakeClient) {
+    this.catalogPulginName = inputName;
+    this.catalogProperties = ImmutableMap.copyOf(properties);
+    this.projectId = propProjectId;
+    this.location = propLocation;
+    Preconditions.checkNotNull(bigLakeClient, "BigLake client must not be null");
+    this.client = bigLakeClient;
+
+    // Users can specify the BigLake catalog ID, otherwise catalog plugin will be used.
+    this.catalogId = properties.getOrDefault(PROPERTIES_KEY_BLMS_CATALOG, inputName);
+    this.catalogName = CatalogName.of(projectId, location, catalogId);
+    LOG.info("Use BigLake catalog: {}", catalogName.toString());
+
+    String fileIOImpl =
+        properties.getOrDefault(CatalogProperties.FILE_IO_IMPL, ResolvingFileIO.class.getName());
+    this.fileIO = CatalogUtil.loadFileIO(fileIOImpl, properties, conf);
+  }
+
+  @Override
+  protected TableOperations newTableOps(TableIdentifier identifier) {
+    // The identifier of metadata tables is like "ns.table.files".
+    // We return a non-existing table in this case (empty table ID is disallowed in BigLake
+    // Metastore), loadTable will try loadMetadataTable.
+    if (identifier.namespace().levels().length > 1
+        && MetadataTableType.from(identifier.name()) != null) {
+      return new BigLakeTableOperations(
+          client, fileIO, getTableName(identifier.namespace().level(0), /* tableId= */ ""));
```

Comment:
Passing an empty string is always suspicious. I think the idea is to get `BigLakeTableOperations` to throw `NoSuchTableException` later after checking the namespace here.

I think it's a better pattern to override `isValidIdentifier` so that this check is done before calling `newTableOps`.

Reply:
Used isValidIdentifier instead, thanks!